### PR TITLE
8309880: Add support for linking libffi on Windows and Mac

### PR DESCRIPTION
**Pull request by JornVernee**
Update the make/autoconf/lib-ffi.m4 script to support using libffi on Windows and Mac.

For Windows I had to tweak `fallbackLinker.c` to be able to build: there was an import of `stdint.h` missing, and since it was using `WSAGetLastError` it needed to link against `ws2_32.lib`.

This PR also contains a fix originally made by @shipilev as part  of: https://github.com/openjdk/jdk/pull/13827 (with a minor tweak), in order to be able to build the fallback linker on mac correctly. I also disabled 3 tests that were failing when using the (libffi-based) fallback linker on that platform.

---

I've updated the `createLibffiBundle.sh` script for Mac, since I got it to work with a devkit, but I didn't manage to do the same for Windows. The steps I took to make my Windows libffi bundle were as follows:
1. run 'x64 Native Tools Command Prompt for VS 2022'. cl.exe and link.exe should be on path
2. run `ucrt64` (this is one of the shell environments that comes with MSYS2). This should carry over the environment set up by the VS dev prompt.
3. then, in the libffi repo root folder:
  3.a run `autogen.sh`
  3.b run:
```
bash configure \
  CC="/path/to/libffi/msvcc.sh -m64" \
  CXX="/path/to/libffi/msvcc.sh -m64" \
  CPPFLAGS="-DFFI_BUILDING_DLL" \
  --disable-docs \
  --prefix=<install dest>
```
(`<install dest>` can be whatever you like. That's what you point `--with-libffi` to).

4. run `make install`. This should create the `<intstall dest>` directory with the files: `include/ffi.h`, `include/ffitarget.h`, `lib/libffi.dll`. It also creates a `lib/libffi.lib` file, but it is of the wrong file type, `DLL` rather than `LIBRARY`.
5. Manually create a working `.lib` file:
  5.a use `dumpbin /exports libffi.dll` to get a list of exported symbols
  5.b put them in a `libffi.def` file: `EXPORTS` on the first line, then a symbol on each line following
  5.c run `lib /def:libffi.def /machine:x64 /out:libffi.lib` to create the right `.lib` file (`lib` is a visual studio tool)

---

Testing:
- manual testing on Windows/x64 and Mac/AArch64, by running the `jdk_foreign` test suite with `-Djdk.internal.foreign.CABI=FALLBACK` (i.e. using the fallback linker).
- Linux/x64 Zero test run of the `jdk_foreign` suite
- Linux/AArch64 Zero build

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309880](https://bugs.openjdk.org/browse/JDK-8309880): Add support for linking libffi on Windows and Mac (**Enhancement** - P4)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Contributors
 * Aleksey Shipilev `<shade@openjdk.org>`
 * Jorn Vernee `<jvernee@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14446/head:pull/14446` \
`$ git checkout pull/14446`

Update a local copy of the PR: \
`$ git checkout pull/14446` \
`$ git pull https://git.openjdk.org/jdk.git pull/14446/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14446`

View PR using the GUI difftool: \
`$ git pr show -t 14446`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14446.diff">https://git.openjdk.org/jdk/pull/14446.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14446#issuecomment-1589364753)

### CODE REVIEW

File: make/modules/java.base/Lib.gmk [hunk 1]
```diff
@@ -228,6 +228,7 @@ ifeq ($(ENABLE_FALLBACK_LINKER), true)
       LDFLAGS := $(LDFLAGS_JDKLIB) \
                  $(call SET_SHARED_LIBRARY_ORIGIN), \
       LIBS := $(LIBFFI_LIBS), \
+	  LIBS_windows := $(LIBFFI_LIBS) ws2_32.lib, \
```
Comment (erikj79) [inline make/modules/java.base/Lib.gmk:231]: Indentation looks off. Make sure you aren't mixing in tabs here.

**Review by erikj79 — COMMENTED** [review]
Could you add the instructions for building libffi for Windows in a comment in createLibffiBundle.sh? I think it would be good to keep those around now that you have figured out how to do it.

File: make/modules/java.base/Lib.gmk (hunk 1 above)
Comment (JornVernee) [reply make/modules/java.base/Lib.gmk:231]: Good catch, I indeed had a tab in there (I have github set to show tabs as 4 spaces, so didn't notice this).

**Review by erikj79 — APPROVED** [review]

**Pull request closed without merging**